openhcl/underhill_attestation/src/lib.rs

```diff
@@ -1063,3 +1063,675 @@ async fn persist_all_key_protectors(
+        .unwrap();
+
+        // Since both `should_write_kp` and `use_gsp_by_id` are true, both key protectors should be updated
+        let found_key_protector = vmgs::read_key_protector(&mut vmgs, AES_WRAPPED_AES_KEY_LENGTH)
+            .await
+            .unwrap();
+        assert_eq!(found_key_protector.as_bytes(), key_protector.as_bytes());
+
+        let found_key_protector_by_id = vmgs::read_key_protector_by_id(&mut vmgs).await.unwrap();
+        assert_eq!(
+            found_key_protector_by_id.as_bytes(),
+            key_protector_by_id.inner.as_bytes()
+        );
+    }
+
+    #[async_test]
+    async fn unlock_vmgs_with_derived_keys() {
+        let mut vmgs = new_formatted_vmgs().await;
+
+        let mut key_protector = new_key_protector();
+        let mut key_protector_by_id = new_key_protector_by_id(None, None, false);
+
+        let ingress = [1; AES_GCM_KEY_LENGTH];
+
+        let derived_keys = Keys {
+            ingress,
+            egress: [2; AES_GCM_KEY_LENGTH],
+        };
+
+        vmgs.add_new_encryption_key(&ingress, EncryptionAlgorithm::AES_GCM)
+            .await
+            .unwrap();
+
+        let key_protector_settings = KeyProtectorSettings {
+            should_write_kp: true,
+            use_gsp_by_id: true,
+            use_hardware_unlock: false,
+        };
+
+        let bios_guid = Guid::new_random();
+
+        unlock_vmgs_data_store(
+            &mut vmgs,
+            true,
+            &mut key_protector,
+            &mut key_protector_by_id,
+            Some(derived_keys),
+            key_protector_settings,
+            bios_guid,
+        )
+        .await
+        .unwrap();
+
+        // Since both `should_write_kp` and `use_gsp_by_id` are true, both key protectors should be updated
+        let found_key_protector = vmgs::read_key_protector(&mut vmgs, AES_WRAPPED_AES_KEY_LENGTH)
+            .await
+            .unwrap();
+        assert_eq!(found_key_protector.as_bytes(), key_protector.as_bytes());
+
+        let found_key_protector_by_id = vmgs::read_key_protector_by_id(&mut vmgs).await.unwrap();
+        assert_eq!(
+            found_key_protector_by_id.as_bytes(),
+            key_protector_by_id.inner.as_bytes()
+        );
+    }
+
+    #[async_test]
+    async fn unlock_vmgs_with_derived_keys_with_rotation() {
+        let mut vmgs = new_formatted_vmgs().await;
+
+        let mut key_protector = new_key_protector();
+        let mut key_protector_by_id = new_key_protector_by_id(None, None, false);
+
+        let additional_key = [2; AES_GCM_KEY_LENGTH];
+        let egress = [3; AES_GCM_KEY_LENGTH];
+
+        let derived_keys = Keys {
+            ingress: [1; AES_GCM_KEY_LENGTH],
+            egress,
+        };
+
+        let additional_key_index = vmgs
+            .add_new_encryption_key(&additional_key, EncryptionAlgorithm::AES_GCM)
+            .await
+            .unwrap();
+        assert_eq!(additional_key_index, 0);
+
+        let egress_key_index = vmgs
```

Comment:
This test seems to cover the case where VMGS isn't locked with ingress but with egress, so `unlock_vmgs_data_store` still succeed (we seem to fall back to egress when attempt with ingress failed) and no key rotation actually occurs. Can you double check and update the name to reflect the scenario?
Can you also add a test that covers the normal rotation case? Like encrypting vmgs and `ingress`. After `unlock_vmgs_data_store`, `unlock_with_encryption_key` with `egress` succeeds but with `ingress` fails. You can also checking the index like you do here.

Reply:
The latest update does a better job of validating lock/unlock assumptions and renames tests to be clearer. Let me know if the latest changes address your comments